### PR TITLE
modified get project id

### PR DESCRIPTION
**Pull request by letusfly85**
@Bugagazavr 

I use GitLab (v7.13.0) and drone 0.4-database branch on 28.07.2015.
(drone's https://github.com/drone/drone/commit/4ab17dd9d9ee0ffd13284ae3e09c44b3573e1b02)

I cannot get repository info below

```Go
id := ns(owner, name)

repo_, err := client.Project(path)
```

I modified using below.

https://github.com/Bugagazavr/go-gitlab-client/pull/9

please let me know your thought.

### CODE REVIEW

File: pkg/remote/builtin/gitlab/gitlab.go [hunk 1]
```diff
@@ -164,16 +178,24 @@ func (r *Gitlab) Netrc(u *common.User) (*common.Netrc, error) {
 // a Public Deploy key, if applicable.
 func (r *Gitlab) Activate(user *common.User, repo *common.Repo, k *common.Keypair, link string) error {
 	var client = NewClient(r.URL, user.Token, r.SkipVerify)
-	var path = ns(repo.Owner, repo.Name)
-	var title, err = GetKeyTitle(link)
+
+	println(link)
```
Comment (Bugagazavr) [inline pkg/remote/builtin/gitlab/gitlab.go]: i think this need remove

File: pkg/remote/builtin/gitlab/gitlab.go [hunk 2]
```diff
@@ -164,16 +178,24 @@ func (r *Gitlab) Netrc(u *common.User) (*common.Netrc, error) {
...
+	projectId, err := client.SearchProjectId(repo.Owner, repo.Name)
+	if err != nil || projectId == 0 {
+		return err
+	}
+	id := strconv.Itoa(projectId)
+
+	title, err := GetKeyTitle(link)
 	if err != nil {
+		println(err.Error())
```
Comment (Bugagazavr) [inline pkg/remote/builtin/gitlab/gitlab.go]: And this

**Comment by Bugagazavr** [comment]
im not a fully agreed with this solution, because global search uses `WHERE 'column' LIKE %query%`, for example, if you have a two projects in the same namespace

example:
`my_awesome_namespace/awesome_site` and `my_awesome_namespace/awesome_devops` you can got devops instead site, or site instead devops

You need correct `SearchProjectId` method

```go
for _, project := range projects {
	if project.Namespace.Name == namespace && strings.ToLower(project.Name) == strings.ToLower(name) {
		id = project.Id
	}
}
```

**Comment by Bugagazavr** [comment]
can you show your project namepsace, and name, i will try to reproduce this error

File: pkg/remote/builtin/gitlab/gitlab.go (hunk 1 above)
Comment (letusfly85) [inline pkg/remote/builtin/gitlab/gitlab.go]: I removed, thank you.

File: pkg/remote/builtin/gitlab/gitlab.go (hunk 2 above)
Comment (letusfly85) [inline pkg/remote/builtin/gitlab/gitlab.go]: I removed, thank you.

**Comment by letusfly85** [comment]
@Bugagazavr 

Thank you for replying, we annoyed from such namespace and name combinations.

examples.

```
examples: List(Tuple(Namespace, Name)) = List( ('hoge', 'gitlab-cookbook'), ('App', 'knowhows') )
```

they include characters such as `-` and capitalized character.

I generate below, 
https://github.com/Bugagazavr/go-gitlab-client/pull/10

thank you.

**Comment by letusfly85** [comment]
And we also use `relative_path` with GitLab such as, `https://my.company.domain.com/gitlab/hoge/gitlab-cookbook.git`.

**Comment by Bugagazavr** [comment]
thats really strange can you show me your drone log and gitlab log when you try to create repository in drone with name `hoge/gitlab-cookbook`

**Comment by letusfly85** [comment]
@Bugagazavr 

I got below log by drone by `bin/drone --degub --config=drone.toml`

```Go
[GIN] 2015/07/30 - 00:09:57 | 404 |  234.801724ms | 10.42.5.10:8312 |   POST    /api/repos/hoge/gitlab-cookbook
Error #01: *Gitlab.buildAndExecRequestRaw failed: <404> https://my.gitlab.com/gitlab/api/v3/projects/hoge%2Fgitlab-cookbook
     Meta: Operation aborted
     Error #02: *Gitlab.buildAndExecRequestRaw failed: <404> https://my.gitlab.com/gitlab/api/v3/projects/hoge%2Fgitlab-cookbook
          Meta: Operation aborted
```


We don't use gitlab-ctl, We provisioned manually from source.
So, I checked by seeing `$GITLAB_HOME/log/production.log`, loglevel `info`, there was no request from drone...

**Comment by Bugagazavr** [comment]
you use apache as fronted server?

this uses fulltext search, and this can be really slow if your database is large and contains many projects.

I think, we can use some custom option, like `DRONE_GITLAB_SEARCH`, to use fulltext search, by default we need to use direct API, to work with project

@bradrydzewski what do you think?

**Comment by letusfly85** [comment]
@Bugagazavr 

>you use apache as fronted server?

yes, and some organizes use it, because there are many resources which use apache server for frontend.


>this uses fulltext search, and this can be really slow if your database is large and contains many projects.

We haven't know this...... Is there any document about this? We want to know correctly :worried: 

Thank you.

**Comment by Bugagazavr** [comment]
> yes, and some other organizes which we know also use apache as front server, because there are many resources or experience which use apache server for frontend.

i know but it's a part of apache configuration when hex `%2F` decoded to `/` symbol, this can be fixed with `AllowEncodedSlashes`

> We haven't know this...... Is there any document about this? We want to know correctly

for me full text search takes over 500 ms., when im not need this because i can directly call API without search ( gitlab omnibus with nginx ). This can be used but as not default option, to get it worked without edit apache config.

**Comment by letusfly85** [comment]
@Bugagazavr @bradrydzewski 

>i know but it's a part of apache configuration when hex %2F decoded to / symbol, this can be fixed with AllowEncodedSlashes

Thanks, I will try.

And, I think there are two ways.

* we serve some documents or wikis, issues (it's easy to understand and find out) for drone.io users (because many users don't know and cannot reach to apache configuration..... )

* we modify `gitlab.go` which can apply to full text search behind apache (it not configured).

I prefer to modifying, because many first users won't see some issues and wikis, it is loss of chances for drone.io. I think.

Thank you.

**Comment by letusfly85** [comment]
@Bugagazavr 

>for me full text search takes over 500 ms.

Thanks so much for this info :smile: , I will try to user nginx.
However, it's little difficult to change apache2 to nginx suddenly in our organization. I will wait for a chance.

**Comment by Bugagazavr** [comment]
@letusfly85 

Looks like documentation placed here https://github.com/drone/drone/blob/0.4-database/doc/setup/gitlab.md

im not an expert in apache, it will be good, if someone provide worked apache configuration example with `AllowEncodedSlashes`

Also this PR can be modified to:
```go
        var id string
        if (r.Search) {
	        projectId, err := client.SearchProjectId(owner, name)
	        if err != nil || projectId == 0 {
		        return nil, err
	        }
	        id = strconv.Itoa(projectId)
        } else {
                id = ns(repo.Owner, repo.Name)
        }

```

And description about `SEARCH` option can be putted to docs, with apache case

**Comment by letusfly85** [comment]
@Bugagazavr 

I saw doc, we also can add apache options here. Thank you.

>Also this PR can be modified to:

Thank you, if I use this scope, I want to make function in `pkg/remote/builtin/gitlab/helper.go`.

I tried now, I cannot find out `r.Search`, 
It seems that `r *Gitlab` don't have `Seach` method, what do you want to point?

**Comment by Bugagazavr** [comment]
@letusfly85 

> It seems that r *Gitlab don't have Seach method, what do you want to point?
yep, you need add this option to `config.go`

**Comment by letusfly85** [comment]
@Bugagazavr 

I added `Search` for Gitlab on `config.go`.
And modified `gitlab/gitlab.go` and `gitlab/helper.go`.

Please see it.

**Comment by Bugagazavr** [comment]
@bradrydzewski LGTM

**Comment by bradrydzewski** [comment]
what was the result of testing `AllowEncodedSlashes` in the apache configuration?

**Comment by letusfly85** [comment]
@bradrydzewski 

thank you for replying, we tried our staging environment.

And we got below on drone debug log.

```Go
[GIN] 2015/08/01 - 07:21:50 | 200 |    1.128793ms | 10.*.*.*:2239 |   GET     /
[GIN] 2015/08/01 - 07:21:51 | 200 |     519.515us | 10.*.*.*:2270 |   GET     /api/user
[GIN] 2015/08/01 - 07:21:51 | 200 |     670.084us | 10.*.*.*:2272 |   GET     /api/user/repos
[GIN] 2015/08/01 - 07:22:44 | 404 | 15.230345383s | 10.*.*.*:2803 |   POST    /api/repos/hoge/bundle-training
Error #01: invalid character '<' looking for beginning of value
     Meta: Operation aborted
Error #02: invalid character '<' looking for beginning of value
     Meta: Operation aborted
```

apache version is:

```
Server version: Apache/2.2.14 (Ubuntu)
```

and in our production environment, we user apache server for other applications.... so if we can, we don't want to add more configuration to apache...(because `AllowEncodedSlashes` cannot be used in the `Location` scope :worried: )

**Comment by bradrydzewski** [comment]
does the gitlab team have a recommended fix or configuration? it may be worth researching before adding a workaround to the drone source code

**Comment by Bugagazavr** [comment]
@bradrydzewski 

config example https://gitlab.com/gitlab-org/gitlab-recipes/blob/73abde7501b6cd128ce22ea9f82dc19fdc867c80/web-server/apache/gitlab-apache2.4.conf
gitlab team recommend to use omnibus https://github.com/gitlabhq/gitlabhq/blob/master/doc/install/installation.md#consider-the-omnibus-package-installation

**Comment by bradrydzewski** [comment]
@Bugagazavr thanks for the reference configuration files.

@letusfly85 any thoughts on why the recommended settings in the GitLab apache configuration file don't work? Do we need to log an issue with the GitLab team about it?

```
  # Ensure that encoded slashes are not decoded but left in their encoded state.
  # http://doc.gitlab.com/ce/api/projects.html#get-single-project
  AllowEncodedSlashes NoDecode
```

**Comment by letusfly85** [comment]
@bradrydzewski 

we cannot use `NoDecode` option on `Apache/2.2.14 `,  maybe this is cause....

http://www.cs.washington.edu/manual/en/mod/core.html#allowencodedslashes

>Do we need to log an issue with the GitLab team about it?

you're right, I will log.

**Comment by letusfly85** [comment]
@bradrydzewski @Bugagazavr 

I asked below.

https://github.com/gitlabhq/gitlabhq/issues/8290

Now, api doesn't handle `namespace/name` correctly in some cases, and GitLab api documents will be fixed for this. (@stanhu if there are mistakes, please let me know)

I will see GitLab source code and try to fix this issue, however, till fixed (by me or someone), some of members cannot use drone (maybe 0.3 also).

Is it reasonable to merge this? And after GitLab is fixed, I will return code.
(or maybe I will use my own branch drone for while...)

**Comment by bradrydzewski** [comment]
@letusfly85 thanks for researching this a bit further. Yes, let's merge this workaround for now. Once it has been fixed upstream in gitlab we can revert back as you've suggested. Thanks @Bugagazavr for helping out here

**Comment by bradrydzewski** [comment]
I also made some minor changes that may need to be ported to GitLab. We are now storing the avatar url in the database instead of the gravatar id. We are also storing the avatar for the repository to accommodate the updated design.

This means setting the avatar has been pushed to the remote plugins. For reference, here is what I did for GitHub:
https://github.com/drone/drone/blob/0.4-database/pkg/remote/builtin/github/github.go#L83
https://github.com/drone/drone/blob/0.4-database/pkg/remote/builtin/github/github.go#L127

**Comment by letusfly85** [comment]
@bradrydzewski 

Thank you!

> Once it has been fixed upstream in gitlab we can revert back as you've suggested.
Yes we will.

@Bugagazavr 

Thank you so much for many advises!

**Comment by letusfly85** [comment]
@bradrydzewski 

Cool, I will see about avator!

**Comment by letusfly85** [comment]
@bradrydzewski 

I saw avatar url for github!

and created PR for gitlab :smile: 

#1123

**Pull request merged**